### PR TITLE
Boolean query

### PR DESCRIPTION
**Pull request by Rythmag**
1. RegexStrings class to define the the query string to be used in regex search
2. Keywords class to define the query string used in match query.
3. Boolean query class to combine match query and regex query and build search request.

### CODE REVIEW

File: src/main/java/com/google/error_analyzer/backend/BooleanQuery.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        MatchQueryBuilder matchQuery = buildMatchQuery();
+        RegexpQueryBuilder regexQuery = buildRegexQuery();
+        BoolQueryBuilder boolQuery = new BoolQueryBuilder()
+            .minimumShouldMatch(1)
+            .should(regexQuery)
+            .should(matchQuery);
+        return boolQuery;
+    }
+
+    private MatchQueryBuilder buildMatchQuery() {
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/backend/BooleanQuery.java:61]: You can create a function : buildMatchQuery(String queryString) {}
Here pass Keywords.getQueryString() as input.

Similarly for buildRegexQuery.

File: src/test/java/com/google/error_analyzer/backend/BooleanQueryTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        BooleanQuery boolQuery = new BooleanQuery();
+        SearchRequest searchRequest = boolQuery.createSearchRequest(fileName);
+        String[] indices = searchRequest.indices();
+        Assert.assertEquals(indices.length, 1);
+        Assert.assertEquals(indices[0], fileName);
+    }
+
+    @Test
+    public void keyWordQueryString() {
+        String actual = Keywords.getQueryString();
```
Comment (shudhatmajain) [inline src/test/java/com/google/error_analyzer/backend/BooleanQueryTest.java:60]: Please test only BooleanQuery functions here.
I think you can test keyWordQueryString & regexpQueryString by checking the values in searchRequest.

File: src/test/java/com/google/error_analyzer/backend/BooleanQueryTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    @Test
+    public void regexpQueryString() {
+        String actual = RegexStrings.getQueryString();
+        String expected = ".*exception";
+        Assert.assertEquals(expected, actual);
+    }
+
+    @Test
+    public void errorFileNameAppend() {
```
Comment (shudhatmajain) [inline src/test/java/com/google/error_analyzer/backend/BooleanQueryTest.java:73]: Please create a separate test file for LogDaoHelper.

File: src/main/java/com/google/error_analyzer/backend/MockErrorQuery.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import com.google.error_analyzer.data.RegexStrings;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/** 
+* This class mocks the BooleanQuery of elasticsearch.
+*/
+
+public class MockErrorQuery {
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/backend/MockErrorQuery.java:24]: Where is ErrorQuery class?

File: src/main/java/com/google/error_analyzer/data/RegexStrings.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        }
+    };
+
+    public static String getQueryString() {
+        String queryString = "";
+        for (String keyword : regexList) {
+            if (queryString.length() == 0) {
+                queryString = queryString.concat(keyword);
+            } else {
+                queryString = queryString.concat("|" + keyword);
```
[the remark this answers is not in the document]
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/data/RegexStrings.java:33]: Both are according to the format required by elasticSearch

File: .gitignore [hunk 6]
```diff
@@ -1,6 +1,10 @@
+
+README*
```
[the remark this answers is not in the document]
Comment (Rythmag) [reply .gitignore:2]: I added this after cloudshell readme was added to git repo

File: src/main/java/com/google/error_analyzer/backend/BooleanQuery.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+
+/** Build search request for boolean query.
+* Match query uses Keywords class for query string
+* Regex query uses RegexStrings class for query string
+*/
+
+public class BooleanQuery {
+    private final String logTextField = LogFields.logTextField;
```
[the remark this answers is not in the document]
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/BooleanQuery.java:34]: These fields are used again for StackTrace.

File: src/test/java/com/google/error_analyzer/backend/BooleanQueryTest.java (hunk 3 above)
Comment (Rythmag) [reply src/test/java/com/google/error_analyzer/backend/BooleanQueryTest.java:73]: There is a LogDaoHelperTest. I'll move this test there

File: src/main/java/com/google/error_analyzer/backend/BooleanQuery.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+
+/** Build search request for boolean query.
+* Match query uses Keywords class for query string
+* Regex query uses RegexStrings class for query string
+*/
+
+public class BooleanQuery {
+    private final Integer requestSize = 10000; //limited by ElasticSearch settings
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/backend/BooleanQuery.java:34]: static

Similarly for below.

File: src/main/java/com/google/error_analyzer/data/constant/Keywords.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+package com.google.error_analyzer.data.constant;
+
+import java.util.ArrayList;
+
+/**
+* This class contains all the keywords used in fulltext query.
+* A log line will apear in search hits if one or more of these terms appear in it.
+*/
+public class Keywords {
+    private final static String LogicString = " OR ";
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/data/constant/Keywords.java:21]: OR_OPERATOR
or
OR

Use CONSTANT_CASE

File: src/main/java/com/google/error_analyzer/data/constant/LogFields.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.*/
+
+package com.google.error_analyzer.data.constant;
+
+/** Field names in an index file.
+*/
+public class LogFields {
+    public static final String logTextField = "logText";
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/data/constant/LogFields.java:17]: LOG_TEXT

Use CONSTANT_CASE.
Can prolly skip "FIELD".

File: src/main/java/com/google/error_analyzer/data/constant/RegexStrings.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+package com.google.error_analyzer.data.constant;
+
+import java.util.ArrayList;
+
+/**
+* This class contains all the regular expressions used in regex query.
+* A log line will apear in search hits if a a term matches the regex.
+*/
+public class RegexStrings{
+    private final static String logicString = "|";
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/data/constant/RegexStrings.java:24]: PIPE_OPERATOR

Use CONSTANT_CASE.

File: src/main/java/com/google/error_analyzer/data/constant/RegexStrings.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    private static final ArrayList < String > regexList = new ArrayList < String > () {
```
Comment (shudhatmajain) [inline src/main/java/com/google/error_analyzer/data/constant/RegexStrings.java:25]: Use CONSTANT_CASE.

File: src/test/java/com/google/error_analyzer/backend/BooleanQueryTest.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+import org.junit.runners.JUnit4;
+import org.junit.Test;
+
+public final class BooleanQueryTest {
+
+    private final String fileName = "file"; 
+    SearchRequest searchRequest = null;
+
+    @Before
+    public void setUp() {
```
Comment (shudhatmajain) [inline src/test/java/com/google/error_analyzer/backend/BooleanQueryTest.java:36]: No need of setup. Create "expected" (which is currently searchRequest) in createSearchRequestWithBoolQuery function itself.

**Pull request merged**